### PR TITLE
Add requirement for input to AllGatherIntoTensor to be contiguous

### PR DESCRIPTION
**Pull request by ani300**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109561

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

### CODE REVIEW

**Review by Chillee — APPROVED** [review]
LGTM, but cc: @wconstab if he has any thoughts on test workaround.

**Comment by ani300** [comment]
I just copied the test workaround from every other test in that file, but @wconstab feel free to chime in

**Comment by kit1980** [comment]
@Chillee @ani300 Let's merge this?

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4790,7 +4790,9 @@ def allreduce(input, reduce_op, tag, ranks, group_size):
     @register_lowering(c10d_functional.all_gather_into_tensor)
     def all_gather_into_tensor(shard, tag, ranks, group_size):
         return TensorBox.create(
-            ir.AllGatherIntoTensor.create(shard, tag, ranks, group_size)
+            ir.AllGatherIntoTensor.create(
+                ir.ExternKernel.require_contiguous(shard), tag, ranks, group_size
```
Comment (wconstab) [inline torch/_inductor/lowering.py:4794]: is this a pure fix or an optimization/tradeoff?  Guess i'm not sure what require_contiguous does exactly. 

Is the issue that previously the input to allgather may have been accessed incorrectly by allgather due to ignoring actual strides?

Comment (Chillee) [reply torch/_inductor/lowering.py:4794]: This is a pure fix.

**Comment by wconstab** [comment]
> I just copied the test workaround from every other test in that file, but @wconstab feel free to chime in

oh sorry- which part is the workaround?

**Comment by ani300** [comment]
@wconstab I think they are referring to 
```
    # TODO: somehow inductor bg compile threads are causing hangs at exit with distributed work dtor
    @patch.object(torch._inductor.config, "compile_threads", 1)
```
as the workaround

**Comment by wconstab** [comment]
oh. yea, this is present on all our distributed/inductor tests, i have no concerns about this (other than it is a long standing issue we should fix, but orthogonal)

**Pull request closed without merging**